provisioning/transport/http/src/http.ts

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+import { EventEmitter } from 'events';
+import { RestApiClient, Http as Base } from 'azure-iot-http-base';
+import { errors } from 'azure-iot-common';
+import * as machina from 'machina';
+import { ClientRequest } from 'http';
+
+export type ProvisioningResponseCallback = (err: Error, result?: any) => void;
+
+export class Http extends EventEmitter {
+
+  private _idScope: string;
+  private _registrationId: string;
+  private _body: any;
+  private _forceRegistration: boolean;
+  private _operationId: string;
+
+  private _fsm: machina.Fsm;
+  private _apiVersion: string = '2016-11-14';
+  private _provisioningServer: string = 'https://global.azure-devices-provisioning.net';
+  private _httpTimeout: number = 4000;
+  private _operationStatusPollingInterval: number = 5000;
+  private _pollingTimer: any;
+  private _currentHttpRequest: ClientRequest;
+  private _restApiClient: RestApiClient;
+
+  // TODO: how to get the user agent string into here?
+
+  /**
+   * @private
+   * @constructor
+   * @param config The configuration object.
+   */
+  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_001: [ The `Http` constructor shall accept the following properties:
+  - `idScope` - a string specifiying the scope of the provisioning operations,
+  - `registrationId` - the registration id for the specific device ] */
+  constructor(idScope: string, registrationId: string, httpBase?: Base) {
+    super();
+    this._idScope = idScope;
+    this._registrationId = registrationId;
+
+    this._fsm = new machina.Fsm({
+      namespace: 'http-provisioning',
+      initialState: 'idle',
+      states: {
+        idle: {
+          _onEnter: () => {
+            this._pollingTimer = null;
+            this._currentHttpRequest = null;
+            this._restApiClient = null;
+            this._body = null;
+            this._forceRegistration = false;
+            this._operationId = '';
+          },
+          register: (authorization, forceRegistration, body, callback) => {
+            this._forceRegistration = forceRegistration;
+            this._body = body;
+            this._restApiClient = new RestApiClient({ 'host' : this._provisioningServer , 'sharedAccessSignature' : authorization}, 'azure-node-provisioning', httpBase);
```

Comment:
>azure-node-provisioning [](start = 133, length = 23)

need to land this with PMs and telemetry folks asap - talk with Yi?

Reply:
will take offline

---
In reply to: [141480500](https://github.com/Azure/azure-iot-sdk-node/pull/93#discussion_r141480500) [](ancestors = 141480500)